accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -621,6 +614,488 @@ It is a warning to use a null aware operator (`?.`, `?[]`, `?..`, `??`, `??=`, o
+  } else {
+    return;
+  }  // Ok, z has inferred type `int` even though it is unassigned on one path.
+}
+```
+
+At non-back edge join points, some adjustment to the **initialization inferred
+type** is done to account for nullability.  Informally, a non-`late` **untyped**
+variable must be assigned a value on every path, but a variable declared `late`
+may be assigned a value on only some of the paths into a join.  At a join point,
+the **initialization inferred type** on each path that has one must agree,
+except that they may differ on nullability, in which case they are adjusted as
+described below.  In all cases, the values assigned must all have the same type
+(up to nullability), or type `Null` - otherwise the analysis fails to infer a
+type for the variable and it is an error.  More precisely, the **initialization
+inferred type** of a variable after a join point is specified as follows.
+
+At a non-backed edge join point in the program, if a non-`late` **untyped
+variable** has one or more *reachable* paths into the join on which it has no
+**initialization inferred type**, and one or more *reachable* paths into the
+join on which it has an **initialization inferred type**, it is an error and
+inference fails for the variable. In other words, a non-`late` variable must
+either be assigned a value on every reachable path or on no reachable paths.
+
+At a non-back edge join point in the program, if an **untyped variable** has no
+**initialization inferred type** on every path into the join, then it has no
+**initialization inferred type** after the join.
+
+If neither of the above cases apply, then there is at least one path (reachable
+or not) into the join which has an **initialization inferred type**, and the
+**initialization inferred type**  for the variable after the join point
+is computed as follows:
+  - If any path into the join (reachable or not) has a **final initialization
+    inferred type** `T`, then:
+    - It is an error if any other path (reachable or not) has a **final
+      initialization type** `S` and `S` is not equal to `T`.
+    - It is an error if any other path (reachable or not) has the **provisional
+      initialization type** `Null` and `T` is not nullable.
+    - It is an error if any other path (reachable or not) has a **provisional
+      initialization type** `S` such that `S` is not `Null`, and neither
+      `S` nor `S?` are equal to `T`.
+    - Otherwise, the **final initialization type** after the join is `T`.
+  - Otherwise, if the **provisional initialization inferred type** on each path
+    which has one (reachable or not) is the same type `T` (using syntactic
+    equality), then the **provisional initialization inferred type** after the
+    join is `T`.
+  - Otherwise, if the **provisional initialization inferred type** on each path
+    which has one (reachable or not) is one of `T`, `T?` or `Null` for some type
+    `T` (using syntactic equality), then the **provisional initialization
```

Comment:
I'm concerned about using syntactic equality here rather than normalized equality.  Does that mean that if the provisional initialization inferred type on one of the paths is `Never?`, then this rule doesn't apply?  That seems bad.